File: sdks/python/src/licensecc/config_attestation.py
"""Verify ``lcccfg1`` config-attestation tokens — a port of ``ConfigAttestation.cpp``.

Same envelope/canonical-payload discipline as the online assertion, with a
config-attestation purpose and a ``config-hash`` claim that binds the token to
the exact config bytes. Fail-closed: typed rejection, never a raw exception.
"""

from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass

from ._signed_token import (
    ALG_RSA_PKCS1_SHA256,
    FieldSpec,
    parse_fields_in_order,
    parse_uint64,
    split_envelope,
    verify_payload_signature,
)
from .keys import TrustedPublicKey
from .results import ConfigAttestationClaims, RejectionCode, VerificationResult

ENVELOPE_PREFIX = "lcccfg1"
PURPOSE = "licensecc-config-attestation"
VERSION = "1"
ISSUED_AT_FUTURE_SKEW_SECONDS = 300
MIN_PUBLIC_KEY_BITS = 3072

# Canonical payload field order — must match build_canonical_config_payload.
_FIELDS = [
    FieldSpec("purpose"),
    FieldSpec("version"),
    FieldSpec("alg"),
    FieldSpec("key-id"),
    FieldSpec("project"),
    FieldSpec("feature"),
    FieldSpec("license-fingerprint"),
    FieldSpec("device-hash"),
    FieldSpec("config-id"),
    FieldSpec("config-seq"),
    FieldSpec("config-hash"),
    FieldSpec("issued-at"),
    FieldSpec("expires-at"),
]


@dataclass
class ConfigAttestationExpected:
    """Caller-supplied expectations for a config token.

    ``config_bytes`` are the EXACT bytes the token must attest (its
    ``config-hash`` must equal ``sha256:`` + sha256(config_bytes)).
    ``project``/``feature``/``license_fingerprint``/``device_hash`` are binding
    values. ``min_config_seq`` is the anti-rollback floor. ``now`` defaults to
    wall-clock; pass a fixed value for deterministic tests.
    """

    config_bytes: bytes
    project: str
    feature: str
    license_fingerprint: str
    device_hash: str = ""
    min_config_seq: int = 0
    now: int | None = None


def _parse_canonical_payload(payload_text: str):
    if not payload_text or payload_text[-1] != "\n" or "\r" in payload_text:
        return None, VerificationResult.reject(
            RejectionCode.PAYLOAD_NOT_CANONICAL, "payload is not canonical"
        )
    # Config payload uses validate_values=False (matches the C++ call).
    values, rejection = parse_fields_in_order(payload_text, _FIELDS, validate_values=False)
    if rejection is not None:
        return None, rejection

    config_seq = parse_uint64(values["config-seq"])
    issued_at = parse_uint64(values["issued-at"])
    expires_at = parse_uint64(values["expires-at"])
    if None in (config_seq, issued_at, expires_at):
        return None, VerificationResult.reject(
            RejectionCode.INTEGER_FIELD_MALFORMED, "integer field malformed"
        )

    claims = ConfigAttestationClaims(
        purpose=values["purpose"],
        version=values["version"],
        algorithm=values["alg"],
        key_id=values["key-id"],
        project=values["project"],
        feature=values["feature"],
        license_fingerprint=values["license-fingerprint"],
        device_hash=values["device-hash"],
        config_id=values["config-id"],
        config_seq=config_seq,
        config_hash=values["config-hash"],
        issued_at=issued_at,
        expires_at=expires_at,
    )
    return claims, None


def _validate_claims(claims: ConfigAttestationClaims, expected: ConfigAttestationExpected):
    if (
        claims.purpose != PURPOSE
        or claims.version != VERSION
        or claims.algorithm != ALG_RSA_PKCS1_SHA256
    ):
        return VerificationResult.reject(RejectionCode.METADATA_MISMATCH, "metadata mismatch")
    if (
        claims.project != expected.project
        or claims.feature != expected.feature
        or claims.license_fingerprint != expected.license_fingerprint
        or claims.device_hash != expected.device_hash
    ):
        return VerificationResult.reject(RejectionCode.BINDING_MISMATCH, "request binding mismatch")

    expected_hash = "sha256:" + hashlib.sha256(expected.config_bytes).hexdigest()
    if claims.config_hash != expected_hash:
        return VerificationResult.reject(
            RejectionCode.CONFIG_HASH_MISMATCH, "hash does not match config bytes"
        )

    now = expected.now if expected.now is not None else int(time.time())
    if claims.issued_at > now + ISSUED_AT_FUTURE_SKEW_SECONDS:
        return VerificationResult.reject(RejectionCode.EXPIRED, "issued in the future")
    if claims.expires_at == 0:
        # Never-expiring config tokens are rejected: every config token must
        # carry a finite expiry, matching the C++ verifier.
        return VerificationResult.reject(RejectionCode.NO_EXPIRY, "config token has no expiry")
    if claims.expires_at < claims.issued_at or claims.expires_at < now:
        return VerificationResult.reject(RejectionCode.EXPIRED, "expired")
    if claims.config_seq < expected.min_config_seq:
        return VerificationResult.reject(RejectionCode.ROLLBACK_BELOW_FLOOR, "sequence below the minimum")
    return None


def verify_config_token(
    token: str,
    expected: ConfigAttestationExpected,
    trusted_keys: list[TrustedPublicKey],
) -> VerificationResult:
    """Verify an ``lcccfg1`` token end-to-end. Returns a typed result, never raises."""
    envelope, rejection = split_envelope(token, ENVELOPE_PREFIX)
    if rejection is not None:
        return rejection

    rejection = verify_payload_signature(
        envelope.payload_bytes,
        envelope.payload_text,
        envelope.signature_bytes,
        trusted_keys,
        MIN_PUBLIC_KEY_BITS,
    )
    if rejection is not None:
        return rejection

    claims, rejection = _parse_canonical_payload(envelope.payload_text)
    if rejection is not None:
        return rejection

    rejection = _validate_claims(claims, expected)
    if rejection is not None:
        return rejection

    return VerificationResult.accept(claims)

File: sdks/java/src/main/java/io/licensecc/client/ConfigAttestation.java
package io.licensecc.client;

import java.math.BigInteger;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Fail-closed verification of canonical {@code lcccfg1} configuration attestations. */
public final class ConfigAttestation {
    private static final String PREFIX = "lcccfg1";
    private static final String PURPOSE = "licensecc-config-attestation";
    private static final String VERSION = "1";
    private static final List<String> FIELDS = List.of(
            "purpose", "version", "alg", "key-id", "project", "feature",
            "license-fingerprint", "device-hash", "config-id", "config-seq",
            "config-hash", "issued-at", "expires-at");

    private ConfigAttestation() {}

    public record Expected(byte[] configBytes, String project, String feature,
                           String licenseFingerprint, String deviceHash,
                           BigInteger minConfigSequence, BigInteger now) {
        public Expected {
            configBytes = configBytes == null ? null : configBytes.clone();
            deviceHash = deviceHash == null ? "" : deviceHash;
            minConfigSequence = minConfigSequence == null ? BigInteger.ZERO : minConfigSequence;
        }

        @Override
        public byte[] configBytes() {
            return configBytes == null ? null : configBytes.clone();
        }
    }

    public record Claims(String purpose, String version, String algorithm, String keyId,
                         String project, String feature, String licenseFingerprint,
                         String deviceHash, String configId, BigInteger configSequence,
                         String configHash, BigInteger issuedAt, BigInteger expiresAt) {}

    public static VerificationResult<Claims> verify(String token, Expected expected,
                                                     List<TrustedPublicKey> trustedKeys,
                                                     Set<String> retiredKeyIds) {
        if (expected == null || expected.configBytes() == null) {
            return VerificationResult.reject(RejectionCode.BINDING_MISMATCH, "expected config bytes are required");
        }
        try {
            SignedToken.Envelope envelope = SignedToken.split(token, PREFIX);
            SignedToken.verifySignature(envelope, trustedKeys, 3072, retiredKeyIds);
            Map<String, String> values = SignedToken.parseFields(envelope.payloadText(), FIELDS, false);
            Claims claims = new Claims(values.get("purpose"), values.get("version"), values.get("alg"),
                    values.get("key-id"), values.get("project"), values.get("feature"),
                    values.get("license-fingerprint"), values.get("device-hash"), values.get("config-id"),
                    SignedToken.parseUint64(values.get("config-seq")), values.get("config-hash"),
                    SignedToken.parseUint64(values.get("issued-at")),
                    SignedToken.parseUint64(values.get("expires-at")));
            return validate(claims, expected);
        } catch (SignedToken.Failure failure) {
            return VerificationResult.reject(failure.code(), failure.getMessage());
        } catch (RuntimeException failure) {
            return VerificationResult.reject(RejectionCode.PAYLOAD_NOT_CANONICAL, "token verification failed");
        }
    }

    public static VerificationResult<Claims> verify(String token, Expected expected,
                                                     List<TrustedPublicKey> trustedKeys) {
        return verify(token, expected, trustedKeys, Set.of());
    }

    private static VerificationResult<Claims> validate(Claims claims, Expected expected) {
        if (!PURPOSE.equals(claims.purpose()) || !VERSION.equals(claims.version())
                || !SignedToken.ALGORITHM.equals(claims.algorithm())) {
            return VerificationResult.reject(RejectionCode.METADATA_MISMATCH, "metadata mismatch");
        }
        if (!equal(claims.project(), expected.project()) || !equal(claims.feature(), expected.feature())
                || !equal(claims.licenseFingerprint(), expected.licenseFingerprint())
                || !equal(claims.deviceHash(), expected.deviceHash())) {
            return VerificationResult.reject(RejectionCode.BINDING_MISMATCH, "request binding mismatch");
        }
        if (!SignedToken.sha256(expected.configBytes()).equals(claims.configHash())) {
            return VerificationResult.reject(RejectionCode.CONFIG_HASH_MISMATCH, "hash does not match config bytes");
        }
        BigInteger now = expected.now() == null ? BigInteger.valueOf(Instant.now().getEpochSecond()) : expected.now();
        if (claims.issuedAt().compareTo(now.add(BigInteger.valueOf(300))) > 0) {
            return VerificationResult.reject(RejectionCode.EXPIRED, "issued in the future");
        }
        if (claims.expiresAt().signum() == 0) {
            return VerificationResult.reject(RejectionCode.NO_EXPIRY, "config token has no expiry");
        }
        if (claims.expiresAt().compareTo(claims.issuedAt()) < 0 || claims.expiresAt().compareTo(now) < 0) {
            return VerificationResult.reject(RejectionCode.EXPIRED, "expired");
        }
        if (claims.configSequence().compareTo(expected.minConfigSequence()) < 0) {
            return VerificationResult.reject(RejectionCode.ROLLBACK_BELOW_FLOOR, "sequence below the minimum");
        }
        return VerificationResult.accept(claims, false);
    }

    private static boolean equal(String left, String right) {
        return left != null && left.equals(right);
    }
}
